// material-ui/lib/select-field.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _mixinsStylePropable = require('./mixins/style-propable');

var _mixinsStylePropable2 = _interopRequireDefault(_mixinsStylePropable);

var _textField = require('./text-field');

var _textField2 = _interopRequireDefault(_textField);

var _dropDownMenu = require('./drop-down-menu');

var _dropDownMenu2 = _interopRequireDefault(_dropDownMenu);

var _stylesRawThemesLightRawTheme = require('./styles/raw-themes/light-raw-theme');

var _stylesRawThemesLightRawTheme2 = _interopRequireDefault(_stylesRawThemesLightRawTheme);

var _stylesThemeManager = require('./styles/theme-manager');

var _stylesThemeManager2 = _interopRequireDefault(_stylesThemeManager);

var _mixinsContextPure = require('./mixins/context-pure');

var _mixinsContextPure2 = _interopRequireDefault(_mixinsContextPure);

var SelectField = _react2['default'].createClass({
  displayName: 'SelectField',

  mixins: [_mixinsStylePropable2['default'], _mixinsContextPure2['default']],

  contextTypes: {
    muiTheme: _react2['default'].PropTypes.object
  },

  statics: {
    getChildrenClasses: function getChildrenClasses() {
      return [_textField2['default'], _dropDownMenu2['default']];
    }
  },

  propTypes: {
    autoWidth: _react2['default'].PropTypes.bool,
    children: _react2['default'].PropTypes.node,
    disabled: _react2['default'].PropTypes.bool,
    errorStyle: _react2['default'].PropTypes.object,
    errorText: _react2['default'].PropTypes.node,
    floatingLabelStyle: _react2['default'].PropTypes.object,
    floatingLabelText: _react2['default'].PropTypes.node,
    fullWidth: _react2['default'].PropTypes.bool,
    hintText: _react2['default'].PropTypes.node,
    iconStyle: _react2['default'].PropTypes.object,
    labelMember: _react2['default'].PropTypes.string, //DEPRECATE
    labelStyle: _react2['default'].PropTypes.object, //DEPRECATE
    menuItems: _react2['default'].PropTypes.array, //DEPRECATE
    onBlur: _react2['default'].PropTypes.func,
    onChange: _react2['default'].PropTypes.func,
    onFocus: _react2['default'].PropTypes.func,
    selectFieldRoot: _react2['default'].PropTypes.object,
    selectedIndex: _react2['default'].PropTypes.number, //DEPRECATE

    /**
     * Override the inline-styles of the root element.
     */
    style: _react2['default'].PropTypes.object,
    underlineDisabledStyle: _react2['default'].PropTypes.object,
    underlineFocusStyle: _react2['default'].PropTypes.object,
    underlineStyle: _react2['default'].PropTypes.object,
    value: _react2['default'].PropTypes.any
  },

  //for passing default theme context to children
  childContextTypes: {
    muiTheme: _react2['default'].PropTypes.object
  },

  getChildContext: function getChildContext() {
    return {
      muiTheme: this.state.muiTheme
    };
  },

  getInitialState: function getInitialState() {
    return {
      muiTheme: this.context.muiTheme ? this.context.muiTheme : _stylesThemeManager2['default'].getMuiTheme(_stylesRawThemesLightRawTheme2['default'])
    };
  },

  getDefaultProps: function getDefaultProps() {
    return {
      autoWidth: false,
      fullWidth: false
    };
  },

  //to update theme inside state whenever a new theme is passed down
  //from the parent / owner using context
  componentWillReceiveProps: function componentWillReceiveProps(nextProps, nextContext) {
    var newMuiTheme = nextContext.muiTheme ? nextContext.muiTheme : this.state.muiTheme;
    this.setState({ muiTheme: newMuiTheme });
  },

  getStyles: function getStyles() {
    var floatingLabelText = this.props.floatingLabelText;

    return {
      label: {
        paddingLeft: 0,
        top: floatingLabelText ? 6 : -4
      },
      icon: {
        right: 0,
        top: floatingLabelText ? 22 : 14
      },
      hideDropDownUnderline: {
        borderTop: 'none'
      }
    };
  },

  render: function render() {
    var styles = this.getStyles();
    var _props = this.props;
    var autoWidth = _props.autoWidth;
    var children = _props.children;
    var style = _props.style;
    var labelStyle = _props.labelStyle;
    var iconStyle = _props.iconStyle;
    var underlineDisabledStyle = _props.underlineDisabledStyle;
    var underlineFocusStyle = _props.underlineFocusStyle;
    var underlineStyle = _props.underlineStyle;
    var errorStyle = _props.errorStyle;
    var selectFieldRoot = _props.selectFieldRoot;
    var disabled = _props.disabled;
    var floatingLabelText = _props.floatingLabelText;
    var floatingLabelStyle = _props.floatingLabelStyle;
    var hintText = _props.hintText;
    var fullWidth = _props.fullWidth;
    var errorText = _props.errorText;
    var onFocus = _props.onFocus;
    var onBlur = _props.onBlur;
    var onChange = _props.onChange;
    var value = _props.value;

    var other = _objectWithoutProperties(_props, ['autoWidth', 'children', 'style', 'labelStyle', 'iconStyle', 'underlineDisabledStyle', 'underlineFocusStyle', 'underlineStyle', 'errorStyle', 'selectFieldRoot', 'disabled', 'floatingLabelText', 'floatingLabelStyle', 'hintText', 'fullWidth', 'errorText', 'onFocus', 'onBlur', 'onChange', 'value']);

    return _react2['default'].createElement(
      _textField2['default'],
      {
        style: style,
        floatingLabelText: floatingLabelText,
        floatingLabelStyle: floatingLabelStyle,
        hintText: !hintText && !floatingLabelText ? ' ' : hintText,
        fullWidth: fullWidth,
        errorText: errorText,
        underlineStyle: underlineStyle,
        errorStyle: errorStyle,
        onFocus: onFocus,
        onBlur: onBlur,
        underlineDisabledStyle: underlineDisabledStyle,
        underlineFocusStyle: underlineFocusStyle
      },
      _react2['default'].createElement(
        _dropDownMenu2['default'],
        _extends({
          disabled: disabled,
          style: selectFieldRoot,
          labelStyle: this.mergeStyles(styles.label, labelStyle),
          iconStyle: this.mergeStyles(styles.icon, iconStyle),
          underlineStyle: styles.hideDropDownUnderline,
          autoWidth: autoWidth,
          value: value,
          onChange: onChange
        }, other),
        children
      )
    );
  }
});

exports['default'] = SelectField;
module.exports = exports['default'];
