// fs-cson/lib/index.js
// Generated by CoffeeScript 1.9.2
(function() {
  var readFile, readFileSync, ref, ref1, ref2, ref3, register, updateFile, updateFileSync, writeFile, writeFileSafe, writeFileSafeSync, writeFileSync;

  ref = require('./read_file'), readFile = ref.readFile, readFileSync = ref.readFileSync;

  ref1 = require('./update_file'), updateFile = ref1.updateFile, updateFileSync = ref1.updateFileSync;

  ref2 = require('./write_file'), writeFile = ref2.writeFile, writeFileSync = ref2.writeFileSync;

  ref3 = require('./write_file_safe'), writeFileSafe = ref3.writeFileSafe, writeFileSafeSync = ref3.writeFileSafeSync;

  register = function() {
    return require('./register');
  };

  module.exports = {
    readFile: readFile,
    readFileSync: readFileSync,
    register: register,
    updateFile: updateFile,
    updateFileSync: updateFileSync,
    writeFile: writeFile,
    writeFileSync: writeFileSync,
    writeFileSafe: writeFileSafe,
    writeFileSafeSync: writeFileSafeSync
  };

}).call(this);
